packages/xo-server-sdn-controller/src/index.js

```diff
@@ -166,41 +166,45 @@ class SDNController extends EventEmitter {
     )
 
     // FIXME: we should monitor when xapis are added/removed
-    forOwn(this._xo.getAllXapis(), async xapi => {
-      await xapi.objectsFetched
+    await Promise.all(
+      map(this._xo.getAllXapis(), async xapi => {
+        await xapi.objectsFetched
+        if (this._setControllerNeeded(xapi)) {
+          return
+        }
 
-      if (this._setControllerNeeded(xapi) === false) {
         this._cleaners.push(await this._manageXapi(xapi))
-
         const hosts = filter(xapi.objects.all, { $type: 'host' })
         await Promise.all(
-          map(hosts, async host => {
+          map(hosts, host => {
             this._createOvsdbClient(host)
           })
         )
 
         // Add already existing pool-wide private networks
         const networks = filter(xapi.objects.all, { $type: 'network' })
         forOwn(networks, async network => {
```

Comment:
Same here: `_.forOwn` will not wait for any async execution, here. You might not want to think about it in this PR, though.

Reply:
Yes, another PR is more appropriate.